File: Data Structures & Algorithms/middle-of-the-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow = head
        fast = head

        n = 0

        while fast:
            fast = fast.next
            n += 1
        
        print(slow.val)

        mid = n // 2

        for _ in range(mid):
            slow = slow.next

        return slow

File: Data Structures & Algorithms/reverse-linked-list-ii/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseBetween(ListNode head, int left, int right) {
        ListNode dummy = new ListNode(-1, head);
        ListNode prev = dummy;
        
        for (int i = 0; i < left - 1; i++) {
            prev = prev.next;
        }

        ListNode subHead = prev.next;
        ListNode subTail = subHead;
        for (int i = 0; i < right - left; i++) {
            subTail = subTail.next;
        }

        ListNode nextNode = subTail.next;
        subTail.next = null;
        prev.next = reverseLinkedList(subHead);
        subHead.next = nextNode;

        return dummy.next;
    }

    public ListNode reverseLinkedList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }

        ListNode newHead = reverseLinkedList(head.next);

        head.next.next = head;
        head.next = null;

        return newHead;
    }
}
